### ISSUE TITLE
[Bugs] AMF does not send "Pdu Session Resource Setup Request"

### ISSUE THREAD

**Issue by BraedenDiaz** [free5gc/free5gc#142]
## Describe the bug
AMF does not send "Pdu Session Resource Setup Request" , but it sends "Dl Nas Transport" containing PduSessionEstablishmentAccept. But PduSessionEstablishmentAccept should be sent in "Pdu Session Resource Setup Request".

I think this is not the desired behavior. In this way external gNB like UERANSIM cannot know a PDU session has been established.

## To Reproduce
Use UERANSIM as gNB and attempt to establish a PDU session in it.

## Expected behavior
PduSessionEstablishmentAccept should be sent in "Pdu Session Resource Setup Request" so that gNB knows a PDU session has been established.

## Environment:
 - free5GC Version: v3.0.4
 - OS: Ubuntu 18 64
 - Kernel version: 5.0.0-23-generic
 - go version: 1.14.4 linux/amd64
 - c compiler version (Option): gcc version 7.4.0

### Configuration File

[config.zip](https://github.com/free5gc/free5gc/files/5576796/config.zip)

### PCAP File

[pcap.zip](https://github.com/free5gc/free5gc/files/5576797/pcap.zip)

### Log File

[free5gc-ueransim-log.txt](https://github.com/free5gc/free5gc/files/5576800/free5gc-ueransim-log.txt)

## System architecture

Attempting to use example 2 in the [free5gc wiki] (https://github.com/free5gc/free5gc/wiki/SMF-Config)

**Comment by aligungr** [comment]
Hello free5gc team.

Let me explain this in simpler words.

Now there are two different messages.
1. PDU Session Establishment Accept (NAS message, sent to UE)
2. PDU Session Resource Setup Request (NGAP message, sent to gNB)

According to Braeden's pcap; AMF does send the "PDU Session Establishment Accept" message to the UE. But it does not send PDU Session Resource Setup Request to the gNB.

Therefore, UERANSIM's UE thinks that it has a PDU session, but UERANSIM's gNB does not know anything about the new PDU session, because gNB expects AMF to request from gNB to allocate u-plane resource for a specific PDU session.

---

In short, AMF should send "PDU Session Resource Setup Request" NGAP message to the gNB after successful establishment of a PDU session.

**Comment by BraedenDiaz** [comment]
Thanks @aligungr

**Comment by free5gc-org** [comment]
Hi,
In our opinion, AMF and RAN should perform Initial context setup procedure so that the RAN can establish the necessary overall initial UE context for the UE. So In v3.0.3 and prior version, AMF always send NAS message **Registration Accept** inside the NGAP message **Initial context setup request**. By this way AMF can send NGAP **PDU Session Resource Setup Request** with NAS **PDU Session Establishment Accept**. And we pass the test with Spirent Landslide.
![image](https://user-images.githubusercontent.com/60502770/100737967-dead8680-340f-11eb-91e9-3e77404ac93d.png)


But we got a bug report when test with real GNB & UE from [this post](https://forum.free5gc.org/t/real-ue-register-failed/281).
So in v3.0.4 we fix the issue to support the two different behavior at registration & session establishment procedure.

The following picture shows the first behavior at v3.0.3 and prior version ([free5gc.zip](https://github.com/free5gc/free5gc/files/5622496/free5gc.zip))
![image](https://user-images.githubusercontent.com/60502770/100737967-dead8680-340f-11eb-91e9-3e77404ac93d.png)
The following picture shows the second behavior we support in v3.0.4  ([mate30_3.5_ipv4v6.zip](https://github.com/free5gc/free5gc/files/5622486/mate30_3.5_ipv4v6.zip))
![image](https://user-images.githubusercontent.com/60502770/100737048-9e99d400-340e-11eb-96d5-781c222ea3aa.png)

And we implement this feature by reference to TS 38.413 8.6.1:
AMF will trigger an Initial context setup procedure when *UE Context Request* IE is included in the Initial UE message. And from the pcap file you provided, the RAN doesn't contain this IE, so AMF sent NAS message **Registration Accept** inside the NGAP message **Downlink NAS Transport** and we will trigger initial context setup procedure at the first pdu session establishment (second behavior)

So you should add *UE Context Request* IE to the Initial UE message otherwise you may need to handle the second behavior which will trigger initial context setup procedure at the first pdu session establishment.

**Comment by aligungr** [comment]
@free5gc-org

I think you are absolutely right. There is no problem as long as PDU session resource setup request items are contained in either PDUSessionResourceSetupRequest or InitialContextSetupRequest.

Thanks for the description.

**Comment by free5gc-org** [comment]
This issue has been mentioned on **free5GC**. There might be relevant details there:

https://forum.free5gc.org/t/trying-to-connect-to-commercial-ran-but-initial-context-setup-failure-radionetwork30/598/11

**Comment by RanjithReddy-11** [comment]
> Thanks @aligungr



> Hi,
> In our opinion, AMF and RAN should perform Initial context setup procedure so that the RAN can establish the necessary overall initial UE context for the UE. So In v3.0.3 and prior version, AMF always send NAS message **Registration Accept** inside the NGAP message **Initial context setup request**. By this way AMF can send NGAP **PDU Session Resource Setup Request** with NAS **PDU Session Establishment Accept**. And we pass the test with Spirent Landslide.
> ![image](https://user-images.githubusercontent.com/60502770/100737967-dead8680-340f-11eb-91e9-3e77404ac93d.png)
> 
> But we got a bug report when test with real GNB & UE from [this post](https://forum.free5gc.org/t/real-ue-register-failed/281).
> So in v3.0.4 we fix the issue to support the two different behavior at registration & session establishment procedure.
> 
> The following picture shows the first behavior at v3.0.3 and prior version ([free5gc.zip](https://github.com/free5gc/free5gc/files/5622496/free5gc.zip))
> ![image](https://user-images.githubusercontent.com/60502770/100737967-dead8680-340f-11eb-91e9-3e77404ac93d.png)
> The following picture shows the second behavior we support in v3.0.4 ([mate30_3.5_ipv4v6.zip](https://github.com/free5gc/free5gc/files/5622486/mate30_3.5_ipv4v6.zip))
> ![image](https://user-images.githubusercontent.com/60502770/100737048-9e99d400-340e-11eb-96d5-781c222ea3aa.png)
> 
> And we implement this feature by reference to TS 38.413 8.6.1:
> AMF will trigger an Initial context setup procedure when _UE Context Request_ IE is included in the Initial UE message. And from the pcap file you provided, the RAN doesn't contain this IE, so AMF sent NAS message **Registration Accept** inside the NGAP message **Downlink NAS Transport** and we will trigger initial context setup procedure at the first pdu session establishment (second behavior)
> 
> So you should add _UE Context Request_ IE to the Initial UE message otherwise you may need to handle the second behavior which will trigger initial context setup procedure at the first pdu session establishment.

Hi Bro,

we have similar Kind of Issue where-in we are getting PDU set up request coming from AMF , below is the error snapshot . can you please guide 
![Capture](https://user-images.githubusercontent.com/73252329/110750207-242bec00-8268-11eb-913c-70b4ce949377.PNG)

**Comment by free5gc-org** [comment]
@RanjithReddy-11 

What rel16 5gc are you using?
There is something wrong in your rel16 5gc.
It should reply NGAP PDUSessionResourceSetupRequest with NAS PDUSessionEstablishmentAccept when receiving NAS SessionEstablishmentRequest.
